src/main/java/mezz/jei/ingredients/IngredientAcceptor.java

```diff
@@ -0,0 +1,86 @@
+package mezz.jei.ingredients;
+
+import com.google.common.base.Preconditions;
+import mezz.jei.api.gui.builder.IIngredientAcceptor;
+import mezz.jei.api.ingredients.IIngredientType;
+import mezz.jei.api.ingredients.ITypedIngredient;
+import mezz.jei.api.runtime.IIngredientManager;
+import mezz.jei.util.ErrorUtil;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnmodifiableView;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class IngredientAcceptor implements IIngredientAcceptor<IngredientAcceptor> {
+	private final IIngredientManager ingredientManager;
+	private final List<Optional<ITypedIngredient<?>>> ingredients = new ArrayList<>();
+	private final Set<IIngredientType<?>> types = new HashSet<>();
+
+	public IngredientAcceptor(IIngredientManager ingredientManager) {
+		this.ingredientManager = ingredientManager;
+	}
+
+	@Override
+	public IngredientAcceptor addIngredientsUntyped(List<?> ingredients) {
+		Preconditions.checkNotNull(ingredients, "ingredients");
+
+		for (Object ingredient : ingredients) {
+			Optional<ITypedIngredient<?>> typedIngredient = TypedIngredient.create(this.ingredientManager, ingredient);
+			typedIngredient.ifPresent(i -> this.types.add(i.getType()));
+
+			this.ingredients.add(typedIngredient);
```

Comment:
It feels like adding to this if `typeIngredient` isn't present and is an empty optional is an oversight.

Reply:
`this.ingredients` holds `Optional<ITypedIngredient<?>>` because lists of displayed ingredients are allowed to contain "blank" ingredients in the rotation (displaying nothing). I represent that here with `Optional.empty()` when it used to be a very sneaky `null`.
I'll add some clarifying doc comments to `this.ingredients`

Reply:
Ah okay, that makes more sense.